### PR TITLE
[AMORO-2819]Spark cannot execute the "alter table set identifier field" command on tables in Iceberg format in unified catalog

### LINKED ISSUE #2819: [Bug][SPARK]:  Spark cannot execute the "alter table set identifier field" command on tables in Iceberg format.
### What happened?

enviroment:

- iceberg_test.set_test_pk  is an native iceberg table.
- use SparkUnifiedSessionCatalog as default catalog.

Failed to execute set identifier fields command
```
spark-sql> ALTER TABLE iceberg_test.set_test_pk SET IDENTIFIER FIELDS id;
ALTER TABLE iceberg_test.set_test_pk SET IDENTIFIER FIELDS id
ANTLR Tool version 4.9.3 used for code generation does not match the current runtime version 4.8ANTLR Runtime version 4.9.3 used for parser compilation does not match the current runtime version 4.8ANTLR Tool version 4.9.3 used for code generation does not match the current runtime version 4.8ANTLR Runtime version 4.9.3 used for parser compilation does not match the current runtime version 4.8org.apache.spark.SparkException: The Spark SQL phase planning failed with an internal error. Please, fill a bug report in, and provide the full stack trace.
	at org.apache.spark.sql.execution.QueryExecution$.toInternalError(QueryExecution.scala:500)
	at org.apache.spark.sql.execution.QueryExecution$.withInternalError(QueryExecution.scala:512)
	at org.apache.spark.sql.execution.QueryExecution.$anonfun$executePhase$1(QueryExecution.scala:185)
	at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:779)
	at org.apache.spark.sql.execution.QueryExecution.executePhase(QueryExecution.scala:184)
	at org.apache.spark.sql.execution.QueryExecution.sparkPlan$lzycompute(QueryExecution.scala:145)
	at org.apache.spark.sql.execution.QueryExecution.sparkPlan(QueryExecution.scala:138)
	at org.apache.spark.sql.execution.QueryExecution.$anonfun$executedPlan$1(QueryExecution.scala:158)
	at org.apache.spark.sql.catalyst.QueryPlanningTracker.measurePhase(QueryPlanningTracker.scala:111)
	at org.apache.spark.sql.execution.QueryExecution.$anonfun$executePhase$2(QueryExecution.scala:185)
	at org.apache.spark.sql.execution.QueryExecution$.withInternalError(QueryExecution.scala:510)
	at org.apache.spark.sql.execution.QueryExecution.$anonfun$executePhase$1(QueryExecution.scala:185)
	at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:779)
	at org.apache.spark.sql.execution.QueryExecution.executePhase(QueryExecution.scala:184)
	at org.apache.spark.sql.execution.QueryExecution.executedPlan$lzycompute(QueryExecution.scala:158)
	at org.apache.spark.sql.execution.QueryExecution.executedPlan(QueryExecution.scala:151)
	at org.apache.spark.sql.execution.QueryExecution.simpleString(QueryExecution.scala:204)
	at org.apache.spark.sql.execution.QueryExecution.org$apache$spark$sql$execution$QueryExecution$$explainString(QueryExecution.scala:249)
	at org.apache.spark.sql.execution.QueryExecution.explainString(QueryExecution.scala:218)
	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$6(SQLExecution.scala:103)
	at org.apache.spark.sql.execution.SQLExecution$.withSQLConfPropagated(SQLExecution.scala:169)
	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$1(SQLExecution.scala:95)
	at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:779)
	at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:64)
	at org.apache.spark.sql.execution.QueryExecution$$anonfun$eagerlyExecuteCommands$1.applyOrElse(QueryExecution.scala:98)
	at org.apache.spark.sql.execution.QueryExecution$$anonfun$eagerlyExecuteCommands$1.applyOrElse(QueryExecution.scala:94)
	at org.apache.spark.sql.catalyst.trees.TreeNode.$anonfun$transformDownWithPruning$1(TreeNode.scala:584)
	at org.apache.spark.sql.catalyst.trees.CurrentOrigin$.withOrigin(TreeNode.scala:176)
	at org.apache.spark.sql.catalyst.trees.TreeNode.transformDownWithPruning(TreeNode.scala:584)
	at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.org$apache$spark$sql$catalyst$plans$logical$AnalysisHelper$$super$transformDownWithPruning(LogicalPlan.scala:30)
	at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper.transformDownWithPruning(AnalysisHelper.scala:267)
	at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper.transformDownWithPruning$(AnalysisHelper.scala:263)
	at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.transformDownWithPruning(LogicalPlan.scala:30)
	at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.transformDownWithPruning(LogicalPlan.scala:30)
	at org.apache.spark.sql.catalyst.trees.TreeNode.transformDown(TreeNode.scala:560)
	at org.apache.spark.sql.execution.QueryExecution.eagerlyExecuteCommands(QueryExecution.scala:94)
	at org.apache.spark.sql.execution.QueryExecution.commandExecuted$lzycompute(QueryExecution.scala:81)
	at org.apache.spark.sql.execution.QueryExecution.commandExecuted(QueryExecution.scala:79)
	at org.apache.spark.sql.Dataset.<init>(Dataset.scala:220)
	at org.apache.spark.sql.Dataset$.$anonfun$ofRows$2(Dataset.scala:100)
	at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:779)
	at org.apache.spark.sql.Dataset$.ofRows(Dataset.scala:97)
	at org.apache.spark.sql.SparkSession.$anonfun$sql$1(SparkSession.scala:622)
	at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:779)
	at org.apache.spark.sql.SparkSession.sql(SparkSession.scala:617)
	at org.apache.spark.sql.SQLContext.sql(SQLContext.scala:651)
	at org.apache.spark.sql.hive.thriftserver.SparkSQLDriver.run(SparkSQLDriver.scala:67)
	at org.apache.spark.sql.hive.thriftserver.SparkSQLCLIDriver.processCmd(SparkSQLCLIDriver.scala:384)
	at org.apache.spark.sql.hive.thriftserver.SparkSQLCLIDriver.$anonfun$processLine$1(SparkSQLCLIDriver.scala:504)
	at org.apache.spark.sql.hive.thriftserver.SparkSQLCLIDriver.$anonfun$processLine$1$adapted(SparkSQLCLIDriver.scala:498)
	at scala.collection.Iterator.foreach(Iterator.scala:943)
	at scala.collection.Iterator.foreach$(Iterator.scala:943)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1431)
	at scala.collection.IterableLike.foreach(IterableLike.scala:74)
	at scala.collection.IterableLike.foreach$(IterableLike.scala:73)
	at scala.collection.AbstractIterable.foreach(Iterable.scala:56)
	at org.apache.spark.sql.hive.thriftserver.SparkSQLCLIDriver.processLine(SparkSQLCLIDriver.scala:498)
	at org.apache.spark.sql.hive.thriftserver.SparkSQLCLIDriver$.main(SparkSQLCLIDriver.scala:286)
	at org.apache.spark.sql.hive.thriftserver.SparkSQLCLIDriver.main(SparkSQLCLIDriver.scala)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.spark.deploy.JavaMainApplication.start(SparkApplication.scala:52)
	at org.apache.spark.deploy.SparkSubmit.org$apache$spark$deploy$SparkSubmit$$runMain(SparkSubmit.scala:958)
	at org.apache.spark.deploy.SparkSubmit.doRunMain$1(SparkSubmit.scala:180)
	at org.apache.spark.deploy.SparkSubmit.submit(SparkSubmit.scala:203)
	at org.apache.spark.deploy.SparkSubmit.doSubmit(SparkSubmit.scala:90)
	at org.apache.spark.deploy.SparkSubmit$$anon$2.doSubmit(SparkSubmit.scala:1046)
	at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:1055)
	at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala)
Caused by: java.lang.AssertionError: assertion failed: No plan for SetIdentifierFields iceberg_test.set_test_pk (id)

	at scala.Predef$.assert(Predef.scala:223)
	at org.apache.spark.sql.catalyst.planning.QueryPlanner.plan(QueryPlanner.scala:93)
	at org.apache.spark.sql.execution.SparkStrategies.plan(SparkStrategies.scala:69)
	at org.apache.spark.sql.catalyst.planning.QueryPlanner.$anonfun$plan$3(QueryPlanner.scala:78)
	at scala.collection.TraversableOnce$folder$1.apply(TraversableOnce.scala:196)
	at scala.collection.TraversableOnce$folder$1.apply(TraversableOnce.scala:194)
	at scala.collection.Iterator.foreach(Iterator.scala:943)
	at scala.collection.Iterator.foreach$(Iterator.scala:943)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1431)
	at scala.collection.TraversableOnce.foldLeft(TraversableOnce.scala:199)
	at scala.collection.TraversableOnce.foldLeft$(TraversableOnce.scala:192)
	at scala.collection.AbstractIterator.foldLeft(Iterator.scala:1431)
	at org.apache.spark.sql.catalyst.planning.QueryPlanner.$anonfun$plan$2(QueryPlanner.scala:75)
	at scala.collection.Iterator$$anon$11.nextCur(Iterator.scala:486)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:492)
	at org.apache.spark.sql.catalyst.planning.QueryPlanner.plan(QueryPlanner.scala:93)
	at org.apache.spark.sql.execution.SparkStrategies.plan(SparkStrategies.scala:69)
	at org.apache.spark.sql.execution.QueryExecution$.createSparkPlan(QueryExecution.scala:459)
	at org.apache.spark.sql.execution.QueryExecution.$anonfun$sparkPlan$1(QueryExecution.scala:145)
	at org.apache.spark.sql.catalyst.QueryPlanningTracker.measurePhase(QueryPlanningTracker.scala:111)
	at org.apache.spark.sql.execution.QueryExecution.$anonfun$executePhase$2(QueryExecution.scala:185)
	at org.apache.spark.sql.execution.QueryExecution$.withInternalError(QueryExecution.scala:510)
	... 69 more

spark-sql>
```

image

### Affects Versions

master

### What engines are you seeing the problem on?

Spark

### How to reproduce

_No response_

### Relevant log output

_No response_

### Anything else

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by Aireed**
## Why are the changes needed?

Close #2819 .

## Brief change log

format: iceberg
engine: spark
Catalog: UnifiedSparkSessionCatalog.
metastore: hms

-   The cause:  UnifiedSessionCatalog is neither SparkCatalog nor SparkSessionCatalog of Iceberg， so it doesn't  match  catalogAndIdentifier, and can't get the physical plan
image

solution:

1. copy the iceberg ExtendedDataSourceV2Strategy and override ` IcebergCatalogAndIdentifier` as MixedFormatExtendedDataSourceV2Strategy
2.  replace ExtendedDataSourceV2Strategy with  MixedFormatExtendedDataSourceV2Strategy in MixedFormatSparkExtensions

Besides the error in the above statement, I found that other SQL syntax issues may have similar problems, so I have fixed them all together here.
The other solutions have been modified as follows.

- Fix the issue of creating a mixed format table using "create table like".
- Shade out the Iceberg classes called by the Amorok Spark extension.
- The Table catalog in Spark 3.3 has added interface support for time travel. Therefore, it is necessary to construct separate SparkSessionCatalog and sparkCatalog for Spark 3.3 and Spark 3.2
- add some unit test for spark 3.2 & spark 3.3

## How was this patch tested?

- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] Run test locally before making a pull request

## Documentation

- Does this pull request introduce a new feature? (yes / no)
- If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by Aireed** [comment]
cc @baiyangtx  @XBaith  @zhoujinsong   PTAL

File: amoro-mixed-format/amoro-mixed-format-spark/v3.3/amoro-mixed-format-spark-3.3/src/main/java/org/apache/amoro/spark/SparkUnifiedCatalog.java [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+   * @param ident a table identifier
+   * @param timestamp timestamp of the table, which is microseconds since 1970-01-01 00:00:00 UTC
+   * @return the table's metadata
+   * @throws NoSuchTableException If the table doesn't exist or is a view
+   */
+  @Override
+  public Table loadTable(Identifier ident, long timestamp) throws NoSuchTableException {
+    TableCatalog tableCatalog = tableCatalog(TableFormat.ICEBERG);
+    if (tableCatalog == null) {
+      throw new UnsupportedOperationException("Doesn't support iceberg table catalog");
```
Comment (baiyangtx) [inline amoro-mixed-format/amoro-mixed-format-spark/v3.3/amoro-mixed-format-spark-3.3/src/main/java/org/apache/amoro/spark/SparkUnifiedCatalog.java:154]: What means 'Doesn't support iceberg table catalog' ?

File: amoro-mixed-format/amoro-mixed-format-spark/amoro-mixed-format-spark-3-common/src/main/java/org/apache/amoro/spark/SparkUnifiedSessionCatalogBase.java [hunk 2]
```diff
@@ -32,20 +32,20 @@
...
 
-  private final Map<TableFormat, SparkTableFormat> tableFormats = Maps.newConcurrentMap();
+  protected final Map<TableFormat, SparkTableFormat> tableFormats = Maps.newConcurrentMap();
 
   @Override
   protected TableCatalog buildTargetCatalog(String name, CaseInsensitiveStringMap options) {
-    SparkUnifiedCatalog sparkUnifiedCatalog = new SparkUnifiedCatalog();
-    sparkUnifiedCatalog.initialize(name, options);
+    SparkUnifiedCatalogBase sparkUnifiedCatalogBase = new SparkUnifiedCatalogBase();
+    sparkUnifiedCatalogBase.initialize(name, options);
```
[the remark this answers is not in the document]
Comment (Aireed) [reply amoro-mixed-format/amoro-mixed-format-spark/amoro-mixed-format-spark-3-common/src/main/java/org/apache/amoro/spark/SparkUnifiedSessionCatalogBase.java:43]: We can remove this function definition from SparkUnifiedSessionCatalogBase and implement it in each specific Spark version. The current implementation doesn't make any sense here.

File: amoro-mixed-format/amoro-mixed-format-spark/v3.2/amoro-mixed-format-spark-3.2/pom.xml [hunk 3]
```diff
@@ -41,6 +41,7 @@
     </properties>
 
     <dependencies>
+
```
[the remark this answers is not in the document]
Comment (Aireed) [reply amoro-mixed-format/amoro-mixed-format-spark/v3.2/amoro-mixed-format-spark-3.2/pom.xml]: removed

File: amoro-mixed-format/amoro-mixed-format-spark/v3.2/amoro-mixed-format-spark-3.2/src/main/java/org/apache/amoro/spark/SparkUnifiedCatalog.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+   * be set as the current namespace.
+   *
+   * <p>The namespace returned by this method is not required to exist.
+   *
+   * @return a multi-part namespace
+   */
+  @Override
+  public String[] defaultNamespace() {
+    return super.defaultNamespace();
+  }
```
[the remark this answers is not in the document]
Comment (Aireed) [reply amoro-mixed-format/amoro-mixed-format-spark/v3.2/amoro-mixed-format-spark-3.2/src/main/java/org/apache/amoro/spark/SparkUnifiedCatalog.java]: removed

File: amoro-mixed-format/amoro-mixed-format-spark/v3.2/amoro-mixed-format-spark-3.2/src/main/java/org/apache/amoro/spark/SparkUnifiedSessionCatalog.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+  @Override
+  protected TableCatalog buildTargetCatalog(String name, CaseInsensitiveStringMap options) {
+    SparkUnifiedCatalog sparkUnifiedCatalog = new SparkUnifiedCatalog();
+    sparkUnifiedCatalog.initialize(name, options);
+    ServiceLoader<SparkTableFormat> sparkTableFormats = ServiceLoader.load(SparkTableFormat.class);
+    for (SparkTableFormat format : sparkTableFormats) {
+      tableFormats.put(format.format(), format);
+    }
+    return sparkUnifiedCatalog;
+  }
```
[the remark this answers is not in the document]
Comment (Aireed) [reply amoro-mixed-format/amoro-mixed-format-spark/v3.2/amoro-mixed-format-spark-3.2/src/main/java/org/apache/amoro/spark/SparkUnifiedSessionCatalog.java:45]: Stay the same as above

File: amoro-mixed-format/amoro-mixed-format-spark/v3.2/amoro-mixed-format-spark-3.2/src/main/java/org/apache/amoro/spark/SparkUnifiedSessionCatalog.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+  @Override
+  public Procedure loadProcedure(Identifier ident) throws NoSuchProcedureException {
+    SparkUnifiedCatalog catalog = (SparkUnifiedCatalog) getTargetCatalog();
+    return catalog.loadProcedure(ident);
+  }
```
[the remark this answers is not in the document]
Comment (Aireed) [reply amoro-mixed-format/amoro-mixed-format-spark/v3.2/amoro-mixed-format-spark-3.2/src/main/java/org/apache/amoro/spark/SparkUnifiedSessionCatalog.java:51]: removed

**Comment by Aireed** [comment]
@baiyangtx I have addressed the review suggestions. Please take another look.

File: amoro-mixed-format/amoro-mixed-format-spark/v3.2/amoro-mixed-format-spark-3.2/src/main/java/org/apache/amoro/spark/SparkUnifiedSessionCatalog.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (baiyangtx) [reply amoro-mixed-format/amoro-mixed-format-spark/v3.2/amoro-mixed-format-spark-3.2/src/main/java/org/apache/amoro/spark/SparkUnifiedSessionCatalog.java:45]:
```suggestion
  // move this method to spark common. 
  protected TableCatalog buildTargetCatalog(String name, CaseInsensitiveStringMap options) {
    SparkUnifiedCatalogBase sparkUnifiedCatalog = createUnifiedCatalog();
    sparkUnifiedCatalog.initialize(name, options);
    ServiceLoader<SparkTableFormat> sparkTableFormats = ServiceLoader.load(SparkTableFormat.class);
    for (SparkTableFormat format : sparkTableFormats) {
      tableFormats.put(format.format(), format);
    }
    return sparkUnifiedCatalog;
  }
  
  // add a abstract create method in base, implement this in 3.2/3.3 module.
  abstract protected SparkUnifiedCatalogBase createUnifiedCatalog();
   
```

File: amoro-mixed-format/amoro-mixed-format-spark/v3.2/amoro-mixed-format-spark-3.2/src/main/java/org/apache/amoro/spark/SparkUnifiedSessionCatalog.java [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+/** @Auth: hzwangtao6 @Time: 2024/5/24 14:04 @Description: */
+public class SparkUnifiedSessionCatalog<
+        T extends TableCatalog & SupportsNamespaces & FunctionCatalog>
+    extends SparkUnifiedSessionCatalogBase<T> {
+
+  @Override
+  protected SparkUnifiedCatalogBase createUnifiedCatalog(
+      String name, CaseInsensitiveStringMap options) {
+    SparkUnifiedCatalog sparkUnifiedCatalog = new SparkUnifiedCatalog();
+    sparkUnifiedCatalog.initialize(name, options);
```
Comment (baiyangtx) [inline amoro-mixed-format/amoro-mixed-format-spark/v3.2/amoro-mixed-format-spark-3.2/src/main/java/org/apache/amoro/spark/SparkUnifiedSessionCatalog.java:38]: Move  call `initialize` to  base.  only create a object is fine.

File: amoro-mixed-format/amoro-mixed-format-spark/v3.3/amoro-mixed-format-spark-3.3/src/main/java/org/apache/amoro/spark/SparkUnifiedCatalog.java [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+  /**
+   * Returns true if the function exists, false otherwise.
+   *
+   * @param ident
+   * @since 3.3.0
+   */
+  @Override
+  public boolean functionExists(Identifier ident) {
+    return FunctionCatalog.super.functionExists(ident);
+  }
```
Comment (baiyangtx) [inline amoro-mixed-format/amoro-mixed-format-spark/v3.3/amoro-mixed-format-spark-3.3/src/main/java/org/apache/amoro/spark/SparkUnifiedCatalog.java:82]: Seems not nessary to override this method.

File: amoro-mixed-format/amoro-mixed-format-spark/v3.3/amoro-mixed-format-spark-3.3/src/main/java/org/apache/amoro/spark/SparkUnifiedSessionCatalog.java [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+
+  private static boolean isSystemNamespace(String[] namespace) {
+    return namespace.length == 1 && namespace[0].equalsIgnoreCase("system");
+  }
+
+  @Override
+  public Procedure loadProcedure(Identifier ident) throws NoSuchProcedureException {
+    SparkUnifiedCatalogBase catalog = (SparkUnifiedCatalogBase) getTargetCatalog();
+    return catalog.loadProcedure(ident);
+  }
```
Comment (baiyangtx) [inline amoro-mixed-format/amoro-mixed-format-spark/v3.3/amoro-mixed-format-spark-3.3/src/main/java/org/apache/amoro/spark/SparkUnifiedSessionCatalog.java:117]: This mehod is looked like same with spark 3.2. could we move it to spark-common?

**Review by baiyangtx — APPROVED** [review]

**Comment by zhoujinsong** [comment]
Thanks for the work! @Aireed 
Thanks for the review! @baiyangtx
